zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/X509ZNodeGroupAclProvider.java

```diff
@@ -0,0 +1,180 @@
+    // Name or Subject Alternative Name from x509 certificate.
+    // The clientId string should be an URI for client and map the client to certain domain.
+    //The user can use the properties defined in X509AuthenticationUtil to extract a desired string as clientId.
+    String clientId;
+    try {
+      clientId = X509AuthenticationUtil.getClientId(clientCert);
+    } catch (Exception e) {
+      // Failed to extract clientId from certificate
+      LOG.error(logStrPrefix + "Failed to extract URI from certificate for session 0x{}",
+          Long.toHexString(cnxn.getSessionId()), e);
+      return KeeperException.Code.OK;
+    }
+
+    // User belongs to super user group
+    if (clientId.equals(System.getProperty(ZOOKEEPER_ZNODEGROUPACL_SUPERUSER))) {
+      cnxn.addAuthInfo(new Id("super", clientId));
+      LOG.info("Authenticated Id '{}' as super user", clientId);
+      return KeeperException.Code.OK;
+    }
+
+    // Get authorized domain names for client
+    Set<String> domains = getUriDomainMappingHelper(serverObjs.getZks()).getDomains(clientId);
+    if (domains.isEmpty()) {
+      // If no domain name is found, use URI as domain name
+      domains = new HashSet<>();
+      domains.add(clientId);
+    }
+
+    Set<String> superUserDomainNames =
+        ZNodeGroupAclProperties.getInstance().getSuperUserDomainNames();
+    for (String domain : domains) {
+      // Grant super user privilege to users belong to super user domains
+      if (superUserDomainNames.contains(domain)) {
+        cnxn.addAuthInfo(new Id("super", clientId));
+        LOG.info(
+            logStrPrefix + "Id '{}' belongs to superUser domain '{}', authenticated as super user",
+            clientId, domain);
+      } else {
+        cnxn.addAuthInfo(new Id(getScheme(), domain));
+        LOG.info(logStrPrefix + "Authenticated Id '{}' for Scheme '{}', Domain '{}'.", clientId,
+            getScheme(), domain);
+      }
+    }
+
+    return KeeperException.Code.OK;
+  }
+
+  @Override
+  public boolean matches(ServerObjs serverObjs, MatchValues matchValues) {
+    for (Id id : serverObjs.getCnxn().getAuthInfo()) {
+      if (id.getId().equals(matchValues.getAclExpr()) || id.getId()
+          .equals(System.getProperty(ZOOKEEPER_ZNODEGROUPACL_SUPERUSER))) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  @Override
+  public String getScheme() {
+    // Same scheme as X509AuthenticationProvider since they both use x509 certificate
+    return "x509";
+  }
+
+  @Override
+  public boolean isAuthenticated() {
+    return true;
+  }
+
+  @Override
+  public boolean isValid(String id) {
+    try {
+      new X500Principal(id);
+      return true;
+    } catch (IllegalArgumentException e) {
+      return false;
+    }
+  }
+
+  private static ClientUriDomainMappingHelper getUriDomainMappingHelper(ZooKeeperServer zks) {
+    if (uriDomainMappingHelper == null) {
+      synchronized (ClientUriDomainMappingHelper.class) {
```

Comment:
why is this synchronizing on the static class object? please synchronize on `uriDomainMappingHelper`.

Reply:
Discussed offline. Decided to synchronize on `this`.